### PR TITLE
chore: Adding VFS testing

### PR DESCRIPTION
**Pull request by yhakbar**
## Description

Starts introducing VFS testing for the `providercache` package.

Seemed like an easy place to start introducing this kind of functionality, as it should have minimal blast radius.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [ ] I authored this code entirely myself
- [x] I am submitting code based on open source software (e.g. MIT, MPL-2.0, Apache)]
- [x] I am adding or upgrading a dependency or adapted code and confirm it has a compatible open source license
- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

## Summary by CodeRabbit

* **Chores**
  * Added a new filesystem dependency and internal virtual-FS utilities to standardize file I/O, ZIP decompression, and archive safety checks.

* **Refactor**
  * Core provider cache, provider service, and CLI config flows now accept an injectable virtual filesystem while retaining OS defaults; initialization paths updated to use the virtual FS.

* **Tests**
  * Test suites migrated to an in-memory filesystem and expanded to cover unzip, symlink, and extraction safety and limits.

---

> [!NOTE]
> **Medium Risk**
> Touches provider caching file I/O and archive extraction paths, which can affect how providers are downloaded/unpacked and cleaned up. While largely an abstraction/testability refactor, the new unzip implementation and FS plumbing could introduce subtle cross-platform or extraction edge cases.
> 
> **Overview**
> Introduces a new `internal/vfs` package (afero-backed) and adds injectable filesystem support across provider caching and CLI config generation so tests can run against an in-memory FS instead of touching the OS.
> 
> Refactors `providercache` and `ProviderService` initialization to use a `NewProviderCache()` builder (`WithFS`) and passes the FS into `cliconfig.LoadUserConfig`, provider service file operations, and local `.terraformrc` creation. Provider archive extraction switches from `go-getter` unzip to a custom `vfs.ZipDecompressor` with ZipSlip/symlink-escape protections plus file-count and total-size limits.
> 
> Updates/extends tests to use the VFS (including homeless env checks) and adds comprehensive unzip/symlink/limit/security test coverage for `internal/vfs`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 36d3a4175986616574d048ed001a6e6151441ec9. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: internal/tf/cache/services/provider_cache.go [hunk 1]
```diff
@@ -375,11 +381,25 @@ func (cache *ProviderCache) acquireLockFile(ctx context.Context) (*util.Lockfile
...
+		ps.fs = fs
+	}
+}
+
 type ProviderService struct {
 	logger                log.Logger
 	providerCacheWarmUpCh chan *ProviderCache
 	credsSource           *cliconfig.CredentialsSource
 
+	// fs is the filesystem for file operations. Defaults to vfs.NewOsFs() if nil.
```
Comment (denis256) [inline internal/tf/cache/services/provider_cache.go]: NewOsFs() or NewOSFS()?

File: internal/tf/cache/services/provider_cache.go [hunk 2]
```diff
@@ -393,7 +413,16 @@ type ProviderService struct {
...
 }
 
-func NewProviderService(cacheDir, userCacheDir string, credsSource *cliconfig.CredentialsSource, logger log.Logger) *ProviderService {
+// getFS returns the configured filesystem or defaults to OsFs.
+func (service *ProviderService) getFS() vfs.FS {
+	if service.fs != nil {
+		return service.fs
+	}
+
+	return vfs.NewOSFS()
```
Comment (denis256) [inline internal/tf/cache/services/provider_cache.go]: Should be initialized `service.fs`? 

It may lead to a memory leak if it is created each time

File: internal/tf/cache/services/provider_cache.go
Comment (denis256) [inline internal/tf/cache/services/provider_cache.go]: How `unzip` function work with vfs? I think we can't inject vfs abstraction in?

File: internal/tf/cache/services/provider_cache.go [hunk 3]
```diff
@@ -273,17 +274,20 @@ func (cache *ProviderCache) setSignature(ctx context.Context) ([]byte, error) {
...
-	if err := os.MkdirAll(filepath.Dir(cache.packageDir), os.ModePerm); err != nil {
+	if err := fs.MkdirAll(filepath.Dir(cache.packageDir), os.ModePerm); err != nil {
 		return errors.New(err)
 	}
 
-	if util.FileExists(cache.userProviderDir) {
+	if vfs.FileExists(fs, cache.userProviderDir) {
 		cache.logger.Debugf("Create symlink file %s to %s", cache.packageDir, cache.userProviderDir)
 
+		// os.Symlink is not VFS-compatible (afero MemMapFs doesn't support symlinks)
```
Comment (denis256) [inline internal/tf/cache/services/provider_cache.go]: Not sure what this comment means - symlink breach VFS abstraction

**Comment by yhakbar** [comment]
@coderabbit review

File: internal/tf/cache/services/provider_cache.go (hunk 3 above)
Comment (yhakbar) [reply internal/tf/cache/services/provider_cache.go]: Added support for symlinks in from afero.

File: internal/tf/cache/services/provider_cache.go
Comment (yhakbar) [reply internal/tf/cache/services/provider_cache.go]: Added unzip functionality into vfs.

File: internal/tf/cache/services/provider_cache.go (hunk 2 above)
Comment (yhakbar) [reply internal/tf/cache/services/provider_cache.go]: Updated everything to use a constructor and reuse the same `service.fs`. It's an empty struct anyways, but it didn't hurt to refactor.

File: internal/tf/cache/services/provider_cache.go (hunk 1 above)
Comment (yhakbar) [reply internal/tf/cache/services/provider_cache.go]: afero uses NewOsFs, and we use NewOSFS, as my understanding is that it's a better practice in Golang.

File: internal/tf/cliconfig/config_test.go [hunk 4]
```diff
@@ -96,23 +87,26 @@ disable_checkpoint_signature = false
...
 
+			// Inject filesystem via options - same Save() method as production
+			tc.config.WithOptions(cliconfig.WithFS(memFs))
+
 			err := tc.config.Save(configFile)
 			require.NoError(t, err)
 
-			hclBytes, err := os.ReadFile(configFile)
+			hclBytes, err := vfs.ReadFile(memFs, configFile)
 			require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (yhakbar) [reply internal/tf/cliconfig/config_test.go:107]: This isn't necessary for `afero.WriteFile`.

File: internal/vfs/vfs.go [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+		return fmt.Errorf("failed to stat zip archive %q: %w", src, err)
+	}
+
+	size := fileInfo.Size()
+
+	var readerAt io.ReaderAt
+	if ra, ok := file.(io.ReaderAt); ok {
+		readerAt = ra
+	} else {
+		data, err := io.ReadAll(file)
```
Comment (denis256) [inline internal/vfs/vfs.go:133]: Large ZIP files it may lead to OOMs

File: internal/vfs/vfs.go [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to create file %q: %w", destPath, err)
+	}
+
+	var reader io.Reader = rc
+
+	if z.FileSizeLimit > 0 {
+		reader = &limitedReader{
+			reader:    rc,
+			remaining: z.FileSizeLimit - *totalSize,
```
Comment (denis256) [inline internal/vfs/vfs.go:309]: No pre-check before extraction of each file, and *totalSize is incremented only after io.Copy succeeds. 

The cumulative total can exceed the limit by up to one buffer (~32KB)

File: internal/vfs/vfs.go [hunk 7]
```diff
@@ -0,0 +1,362 @@
+// Package vfs provides a virtual filesystem abstraction for testing and production use.
+// It wraps afero to provide a consistent interface for filesystem operations.
+package vfs
```
Comment (denis256) [inline internal/vfs/vfs.go:3]: `helpers.FetchToFile()` - bypass usage of VFS

File: internal/vfs/vfs.go (hunk 5 above)
Comment (yhakbar) [reply internal/vfs/vfs.go:133]: That's only gonna be relevant if the reader doesn't implement `io.ReaderAt`, which is only relevant in testing.

File: internal/vfs/vfs.go (hunk 6 above)
Comment (yhakbar) [reply internal/vfs/vfs.go:309]: It's checked within the `io.Copy`:
https://github.com/gruntwork-io/terragrunt/pull/5490/changes#diff-0a7d4945691effa383a7ac883a713cd8f28cb7788be0e1876b07a59e81876596R346

File: internal/vfs/vfs.go (hunk 7 above)
Comment (yhakbar) [reply internal/vfs/vfs.go:3]: Will address that, but it will require a more invasive update into the `tf` package.
